### PR TITLE
Pg hba re comment parsing

### LINKED ISSUE #420: ldapbindpasswd in options which includes  '#' creates faulty pg_hba lines
##### SUMMARY
ldapbindpasswd in options which includes  '#' creates faulty pg_hba lines
While a hashmark in a password might be ugly I cannot change this

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
postgresql_pg_hba

##### ANSIBLE VERSION

```paste below

ansible [core 2.11.12]
  config file = None
  configured module search path = ['/home/USER/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/USER/.local/lib/python3.6/site-packages/ansible
  ansible collection location = /home/USER/.ansible/collections:/usr/share/ansible/collections
  executable location = /home/USER/.local/bin/ansible
  python version = 3.6.8 (default, Aug 13 2020, 07:46:32) [GCC 4.8.5 20150623 (Red Hat 4.8.5-39)]
  jinja version = 3.0.1
  libyaml = True

```

##### COLLECTION VERSION

```paste below
Collection           Version
-------------------- -------
community.postgresql 2.3.2
```

##### CONFIGURATION

```paste below

```

##### STEPS TO REPRODUCE

```yaml
---
- hosts: localhost
  tasks:
    - name: line 1
      community.postgresql.postgresql_pg_hba:
        dest: pg_hba.conf
        create: yes
        users: bug
        databases: bug
        source: 66.66.66.66/32
        contype: hostssl
        method: ldap
        options: ldapbindpasswd ="#BROKEN"
        keep_comments_at_rules: true
        comment: messed up
    - name: line 2 breaks line 1
      community.postgresql.postgresql_pg_hba:
        dest: pg_hba.conf
        create: yes
        users: bug2
        databases: bug
        source: 66.66.66.66/32
        contype: hostssl
        method: ldap
        options: ldapbindpasswd ="#BROKEN"
        keep_comments_at_rules: true
```

##### EXPECTED RESULTS

file pg_hba should be:
```
hostssl bug     bug     66.66.66.66/32  ldap    ldapbindpasswd ="#BROKEN"        #messed up
hostssl bug     bug2    66.66.66.66/32  ldap    ldapbindpasswd ="#BROKEN"
```

##### ACTUAL RESULTS

pg_hba.conf generated is:
```
hostssl bug     bug     66.66.66.66/32  ldap    ldapbindpasswd ="    #BROKEN"        #messed up
hostssl bug     bug2    66.66.66.66/32  ldap    ldapbindpasswd ="#BROKEN"
```

```paste below
PLAY [localhost] *****************************************************************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [line 1] ********************************************************************************************************************************************************************************************************************************************************************************
--- before
+++ after
@@ -1,4 +1,6 @@
 {
     "file": "pg_hba.conf",
-    "pg_hba": []
+    "pg_hba": [
+        "hostssl\tbug\tbug\t66.66.66.66/32\tldap\tldapbindpasswd =\"#BROKEN\""
+    ]
 }

changed: [localhost]

TASK [line 2 breaks line 1] ******************************************************************************************************************************************************************************************************************************************************************
--- before
+++ after
@@ -1,4 +1,6 @@
 {
     "file": "pg_hba.conf",
-    "pg_hba": []
+    "pg_hba": [
+        "hostssl\tbug\tbug2\t66.66.66.66/32\tldap\tldapbindpasswd =\"#BROKEN\""
+    ]
 }

changed: [localhost]

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************************************************************
localhost                  : ok=3    changed=2    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

### PR DESCRIPTION
**Pull request by stelb**
##### SUMMARY

Fixes #420 (# in quoted parameters)

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME
postgresql_pg_hba

##### ADDITIONAL INFORMATION

simple splitting at first hashmark is replaced by a regex

```paste below

```

### CODE REVIEW

**Review by hunleyd — CHANGES_REQUESTED** [review]
add a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment) ? otherwise, lgtm. thank you for making this!

**Comment by stelb** [comment]
ok, I've added a fragment

**Comment by Andersson007** [comment]
@betanummeric would you like to take a look?

**Review by hunleyd — APPROVED** [review]

File: changelogs/fragments/423-postgresql_pg_hba_comment_parsing.yml
Comment (Andersson007) [inline changelogs/fragments/423-postgresql_pg_hba_comment_parsing.yml:2]:
```suggestion
- postgresql_pg_hba - comment parsing assumed any hash starts a comment, but could be in quoted values (https://github.com/ansible-collections/community.postgresql/issues/420).
```
To make all changelog entries consitent.

Let's wait for @betanummeric until tomorrow, then will merge the PR

File: plugins/modules/postgresql_pg_hba.py
Comment (betanummeric) [inline plugins/modules/postgresql_pg_hba.py:369]: This regex does not catch all cases, for example
```
 a "b\"c#d" f # g # h " i
```

suggestion:
```python3
# on module load:
line_comment_regex = re.compile(r'(("(\\"|[^"])*"|[^"#])*)(#(.*))?')

# for every line:
parsed_line = line_comment_regex.fullmatch(line)
line = parsed_line.group(1)
comment = parsed_line.group(5)
```

File: changelogs/fragments/423-postgresql_pg_hba_comment_parsing.yml
Comment (betanummeric) [reply changelogs/fragments/423-postgresql_pg_hba_comment_parsing.yml:2]: The changelog should focus on what changed, this is more of an issue description. Suggestion:
> fix pg_hba.conf parsing: support double-quoted values, including escaped quotes

**Comment by betanummeric** [comment]
The further processing also needs to be fixed to properly support quoted values.

For example, the `PgHbaRule.fromline` method splits the line by whitespace, but a quoted value could include a whitespace. We could probably change that to something like

```python3
line_items = []
for match in re.findall(r'(("(\\"|[^"])*"|[^" ])+)', line):
    match = match[0]
    if match.startswith('"'):
        # remove quotes (do we need this?)
        match = match[1:len(match)-1].replace('\\"', '"')
    line_items.append(match)
```

I have not checked what other pieces of code need to be amended.

**Review by betanummeric — CHANGES_REQUESTED** [review]

File: changelogs/fragments/423-postgresql_pg_hba_comment_parsing.yml
Comment (Andersson007) [reply changelogs/fragments/423-postgresql_pg_hba_comment_parsing.yml:2]: +1 @stelb what do you think ^ ?

Comment (stelb) [reply changelogs/fragments/423-postgresql_pg_hba_comment_parsing.yml:2]: Well, ok. I took a look in one changelog and it seemed to describe the fixed problem too.  I'm fine with your suggestion :)

- All others prefix it by module name
- a colon needs to be in a quoted string in yaml.

```
bugfixes:
  - "postgresql_pg_hba - fix pg_hba.conf parsing: support double-quoted values, including escaped quotes"
```

File: plugins/modules/postgresql_pg_hba.py
Comment (stelb) [reply plugins/modules/postgresql_pg_hba.py:369]: I was afraid, that I still miss a case :-( and there are even more code changes needed..
I'll check again later today

**Comment by Andersson007** [comment]
@stelb any updates on the PR/things we can help with?

**Comment by Andersson007** [comment]
@stelb is there any way we can help?
if you have no time to finish the work, please let us know, maybe someone will take it

File: changelogs/fragments/423-postgresql_pg_hba_comment_parsing.yml
Comment (hunleyd) [inline changelogs/fragments/423-postgresql_pg_hba_comment_parsing.yml:2]:
```suggestion
  - postgresql_pg_hba - add support for double-quoted values and escaped quotes when parsing pg_hba.conf (https://github.com/ansible-collections/community.postgresql/issues/420)
```

**Review by hunleyd — CHANGES_REQUESTED** [review]

**Comment by stelb** [comment]
Hey @Andersson007 
I'm sorry I have no time to finish that  :-(

**Comment by Andersson007** [comment]
> Hey @Andersson007 I'm sorry I have no time to finish that :-(

@stelb no problem at all, thanks for the info! if once you have time, we'd be happy to see contributions from you!
So then i'll ask someone to finish the PR, if you're OK.
Does anyone want to pick it up?

**Comment by stelb** [comment]
I'm fine with that :)
(And maybe I have something else to add/fix, I will prepare that better..)

**Comment by betanummeric** [comment]
I can proceed with this in a new PR. I think the methods `PgHbaRule.line`, `PgHbaRule.fromline` and `PgHba.read` will need changes.

**Comment by Andersson007** [comment]
closing this PR, I'll put a link in the issue that people can use it when fixing the issue, thanks! (if you want to finish it, please reopen the PR)

**Pull request closed without merging**
